.github/workflows/ci.yml

```diff
@@ -42,25 +43,66 @@ jobs:
           version: latest
           args: buildAll
 
-      - name: End to end test
+      - name: Setup services (Trino, Grafana, PostgreSQL, Keycloak)
         run: |
           docker network create trino
 
-          docker run \
-            --rm --detach \
+          echo "Starting PostgreSQL..."
+          docker run --rm --detach \
+            --name postgres \
+            --net trino \
+            --env POSTGRES_USER=keycloak \
+            --env POSTGRES_PASSWORD=keycloak \
+            --env POSTGRES_DB=keycloak \
+            postgres:17.4
+          
+          echo "Starting Keycloak..."
+          docker run --rm --detach \
+            --name keycloak \
+            --net trino \
+            --publish 18080:8080 \
+            --env KC_BOOTSTRAP_ADMIN_USERNAME=admin \
+            --env KC_BOOTSTRAP_ADMIN_PASSWORD=admin \
+            --env KC_DB=postgres \
+            --env KC_DB_URL_HOST=postgres \
+            --env KC_DB_URL_DATABASE=keycloak \
+            --env KC_DB_USERNAME=keycloak \
+            --env KC_DB_PASSWORD=keycloak \
+            --volume "$(pwd)/test-keycloak-realm.json:/opt/keycloak/data/import/realm.json" \
+            quay.io/keycloak/keycloak:26.1.4 \
+            start-dev --import-realm
+
+          echo "Waiting for Keycloak to be ready..."
+          for i in {1..30}; do
+            if curl -s http://localhost:18080/realms/master | grep -q "realm"; then
+              echo "Keycloak is ready!"
+              break
+            fi
+            if [ $i -eq 100 ]; then
```

Comment:
That will never happen if the loop iterates from 1 to 30. Change the loop to a `while true;`?